docs/reactive-nonblocking.qmd

```diff
@@ -0,0 +1,316 @@
+import asyncio
+import concurrent.futures
+import time
+
+from shiny import App, reactive, render, ui
+
+app_ui = ui.page_fluid(
+    ui.input_numeric("x", "x", value=1),
+    ui.input_numeric("y", "y", value=2),
+    ui.input_task_button("btn", "Add numbers"),
+    ui.output_text("sum"),
+)
+
+pool = concurrent.futures.ThreadPoolExecutor()
+
+def slow_sum(x, y):
+    time.sleep(5)  # Simulate a slow synchronous task
+    return x + y
+
+def server(input, output, session):
+    @ui.bind_task_button(button_id="btn")
+    @reactive.extended_task
+    async def sum_values(x, y):
+        loop = asyncio.get_event_loop()
+        return await loop.run_in_executor(pool, slow_sum, x, y)
+
+    @reactive.effect
+    @reactive.event(input.btn)
+    def btn_click():
+        sum_values(input.x(), input.y())
+
+    @render.text
+    def sum():
+        return str(sum_values.result())
+
+app = App(app_ui, server)
+app.on_shutdown(pool.shutdown)
+```
+
+With a small tweak, we can run the task on a different process instead of a different thread.
+This is a good strategy for CPU-bound code, because it allows the task to run concurrently with other Python code.
+
+```{python}
+#| eval: false
+import asyncio
+import concurrent.futures
+import time
+
+from shiny import App, reactive, render, ui
+
+app_ui = ui.page_fluid(
+    ui.input_numeric("x", "x", value=1),
+    ui.input_numeric("y", "y", value=2),
+    ui.input_task_button("btn", "Add numbers"),
+    ui.output_text("sum"),
+)
+
+pool = concurrent.futures.ProcessPoolExecutor()
+
+def slow_sum(x, y):
+    time.sleep(5)  # Simulate a slow synchronous task
+    return x + y
+
+def server(input, output, session):
+    @ui.bind_task_button(button_id="btn")
+    @reactive.extended_task
+    async def sum_values(x, y):
+        loop = asyncio.get_event_loop()
+        return await loop.run_in_executor(pool, slow_sum, x, y)
+
+    @reactive.effect
+    @reactive.event(input.btn)
+    def btn_click():
+        sum_values(input.x(), input.y())
+
+    @render.text
+    def sum():
+        return str(sum_values.result())
+
+app = App(app_ui, server)
+app.on_shutdown(pool.shutdown)
+```
+
+In both of these examples, the `slow_sum` function is defined at the module level, outside of the Shiny server function.
+This is critically important for `ProcessPoolExecutor` to work correctly because of how Python's `multiprocessing` module works: only module-level functions can survive the trip to a worker Python subprocess.
+
+There's also the `app.on_shutdown(pool.shutdown)` line at the end of each example.
+This is necessary to ensure that the pool is shut down when the app stops running.
+In particular, if you're using `ProcessPoolExecutor` and neglect to shut down the pool, you can end up with orphaned Python processes hanging around.
+
+## Summary
+
+Only use async functions in your reactive code if you need to call async-only functions.
+Don't expect your application to run faster, more responsively, or more efficiently.
```

Comment:
```suggestion
Don't expect your application to run faster, more responsively, or more efficiently with async functions alone.
```

Reply:
I like the stronger wording. It's true that async code will perform the same as sync, and we've made it clear there is a different path you need to take if you want concurrency. I think the stronger wording is important because it's such a violation of expectations.

Reply:
👌